{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/bulatych/ML_HSE/blob/main/homeworks/HW_4_(ENCODE).ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "First of all, a few notes:\n",
        "* Getting to know ENCODE might be challenging. If you have not attented the seminar, here are few slides to explain, how to download ENCODE data in the context of this homework: https://github.com/alllirik/hse_ml_bioinf/blob/main/homeworks/HW-4%20Guide.pdf\n",
        "* Check the seventh seminar where we preprocess ENCODE data.\n",
        "* Comment and describe! To give you a high grade, we need to understand that you interpreted all results correctly.\n",
        "\n",
        "Remember to submit the feedback! Especially if the homework was too difficult or easy for you."
      ],
      "metadata": {
        "id": "h6rXJxY5aClz"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Introduction [2.5]\n",
        "\n",
        "This homework is heavily based on the seminar material. Make sure you have it at hand - it should help a lot.\n",
        "\n",
        "Also, it is intentended that you use this data in the next homework."
      ],
      "metadata": {
        "id": "pHJmy6k1bJm2"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Know your protocols\n",
        "\n",
        "During the seminar, we covered basic ideas behind ChIP/ATAC-seq protocols. However, you must understand your data clearly before analyzing it.\n",
        "\n",
        "So here are a few questions:\n",
        "* [0.3] List the main experimental(!) steps of the ChIP-seq protocol, its main advantages, and limitations.\n",
        "\n",
        "CHIP-seq is a method for analyzing DNA-protein interactions based on chromatin immunoprecipitation (ChIP) and highly efficient DNA sequencing.\n",
        "\n",
        "Basic steps:\n",
        "- Formation of reversible crosslinking between DNA and proteins interacting with it\n",
        "- DNA isolation and fragmentation by ultrasound or endonucleases\n",
        "- Deposition of antibodies specific to the protein (Immunoprecipitation)\n",
        "- Destruction of crosslinking between protein and DNA, purification of DNA\n",
        "- Sequencing and analyses\n",
        "\n",
        "Advantages:\n",
        "- single nucleotide resolution.\n",
        "- small amounts are needed for ChIP-seq\n",
        "- no  cross-hybridization between imperfectly matched sequences\n",
        "- high sensitivity , depending on the sequencing depth\n",
        "\n",
        "Limitations:\n",
        "- access and cost\n",
        "- the quality of the data relies on the quality of the antibody\n",
        "- no standart control. The peaks in the profiles generated using ChIP-seq must be compared with the same location in the control sample to determine the significance of the peak.\n",
        "\n",
        "\n",
        "* [0.3] Please, do the same for the ATAC-seq protocol.\n",
        "Steps:\n",
        "- Isolation and lysis of cells\n",
        "- DNA transposition\n",
        "- Library preparation and DNA amplification\n",
        "- Sequencing\n",
        "Advantages:\n",
        "- High sensitivity and resolution (reaches a value of one nucleotide)\n",
        "- Compatibility with different types of samples\n",
        "- Simpe and fast\n",
        "- Allows you to identify various regulatory elements\n",
        "\n",
        "Limitations:\n",
        "- Sequencing artifacts and sensitivity to sample quality\n",
        "- Does not directly determine the binding of transcription factors or other proteins.\n",
        "- The signal may be weak in highly packed chromatin regions\n",
        "\n",
        "* [0.3] Are these experiments universal, or should they be repeated for each culture of interest? Why?\n",
        "\n",
        "→ ATAC-seq and ChIP-seq experiments are not universal and should generally be repeated for each cell type. The pattern of accessible chromatin is highly dynamic and specific to the cell type, developmental stage and so on. Also the binding sites of transcription factors and histone modifications differ depending on the biological context. Obviously repeating experiments validates findings and ensures results are not artifacts of experimental error.\n",
        "\n",
        "* [0.6] Provide a summary of the typical bioinformatic analysis for these assays. (You can/should use information provided in association graph from choosen ENCODE experiment of interest). What is Irreproducibility Discovery Rate (IDR)? What's the difference between regular replicas and pseudoreplicas in terms of IDR?\n",
        "\n",
        "→1. Read Trimming, Concatenation, and Alignment\n",
        "- Quality control of raw sequencing reads. Read Trimming: Low-quality bases and adapters are removed from reads.\n",
        "- Multiple raw sequencing files are merged if necessary\n",
        "- Mapping the cleaned reads to the reference genome\n",
        "2. Signal Generation\n",
        "- Generate a signal track representing the read density over the genome.\n",
        "- Signal is computed as fold-change over control or p-value signal.\n",
        "3. Peak Calling and IDR Analysis\n",
        "- Identify enriched regions (peaks) where significant binding or modification occurs using peak-calling tools (e.g., MACS2).\n",
        "- Splits the sequencing data into pseudoreplicates and performs peak calling independently for each.\n",
        "- Compares pseudoreplicates to calculate the Irreproducibility Discovery Rate (IDR) to assess the consistency of detected peaks.\n",
        "4. File Format Conversion\n",
        "- Convert intermediate output files into standard formats (bigBed to bed)\n",
        "\n",
        "IDR - Irreproducibility Discovery Rate measures consistency between replicates in high-throughput experiments.  Also uses reproducibility in score rankings between peaks in each replicate to determine an optimal cutoff for significance.\n",
        "Peaks from two replicates are ranked based on their significance (e.g., p-value or fold enrichment).\n",
        "IDR measures the proportion of inconsistent peaks between the replicates at various rank thresholds.\n",
        "Lower IDR values indicate higher reproducibility.\n",
        "\n",
        "\n",
        "Regular replicates are biological replicates from independent samples used to assess biological reproducibility with IDR, while pseudoreplicates are artificial subsets of reads from the same experiment used to evaluate technical reproducibility and peak-calling consistency. Together, they validate the reliability of peak-calling and ensure robust identification of meaningful regions"
      ],
      "metadata": {
        "id": "S5r1z_c1bJsY"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Target transcription factors\n",
        "\n",
        "Overall, you need to repeat our seminar work, but this time with more transcriptional factors. I deliberately did not choose any TFs for you, so you can pick your favorite one or just some proteins that look interesting to you.\n",
        "\n",
        "Here is the task:\n",
        "\n",
        "[1] Use ENCODE database and pick a tissue or cell culture with a published ATAC-seq experiment and ***3*** ChIP-seq experiments (for your favorite **transcription factors**). **Ensure all experiments are from the same culture and pass routine ENCODE checks.** Provide links to experiments and descriptions of your TFs below.\n",
        "\n",
        "e.g.\n",
        " * Cell line: K562 - сells were the first human immortalised myelogenous leukemia cell line to be established. K562 cells are of the erythroleukemia type, and the cell line is derived from a 53-year-old female chronic myelogenous leukemia patient in blast crisis.\n",
        " * TF-1: CTCF - is a multifunctional transcription factor and chromatin architectural protein that plays crucial roles in genome organization and gene regulation. CTCF binds to specific DNA sequences called CTCF binding sites, functioning as an insulator to block interactions between enhancers and promoters when placed between them.\n",
        " https://www.encodeproject.org/files/ENCFF942DXM/\n",
        " * TF-2: EP300 It functions as histone acetyltransferase that regulates transcription of genes via chromatin remodeling by allowing histone proteins to wrap DNA less tightly. This enzyme plays an essential role in regulating cell growth and division, prompting cells to mature and assume specialized functions (differentiate), and preventing the growth of cancerous tumors. https://www.encodeproject.org/files/ENCFF449XHW/\n",
        " * TF-3: POLR2A POLR2A (RNA polymerase II subunit A) is the largest catalytic subunit of RNA polymerase II (Pol II), the enzyme responsible for transcribing messenger RNA (mRNA) and various non-coding RNAs in eukaryotes. https://www.encodeproject.org/files/ENCFF927NKT/@@download/ENCFF927NKT.bigBed\n",
        " * ATAC-seq: replica 1: https://www.encodeproject.org/files/ENCFF117MSK/\n",
        " replica 2: https://www.encodeproject.org/files/ENCFF976CEI/\n",
        "\n",
        "\n",
        " →"
      ],
      "metadata": {
        "id": "aiit4jH9dMmz"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Data preprocessing [7.5]"
      ],
      "metadata": {
        "id": "_tGA0Ysjig-t"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "[5]\n",
        "Implement the main workflow in line with classroom material:\n",
        "* Download regions\n",
        "* Calculate intersections / subtractions\n",
        "  * Binding sites for TF can overlap; this is expected. However, here we will use a simplistic worldview and drop such situations. That is, you need to keep and process only specific sites for each TF. If it's not possible - pick a different set of transcription factors.\n",
        "* Get sequences\n",
        "* Calculate k-mers\n",
        "* One-hot encode classes\n",
        "* Download your dataset and save it for the next homework\n"
      ],
      "metadata": {
        "id": "VbKjUd_fikGc"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# 1 ChIP-seq replicas\n",
        "!wget -O CTCF.bigBed  \"https://www.encodeproject.org/files/ENCFF629CUA/@@download/ENCFF629CUA.bigBed\"\n",
        "!wget -O EP300.bigBed \"https://www.encodeproject.org/files/ENCFF671KDT/@@download/ENCFF671KDT.bigBed\"\n",
        "!wget -O POLR2A.bigBed \"https://www.encodeproject.org/files/ENCFF927NKT/@@download/ENCFF927NKT.bigBed\"\n",
        "# 2 ATAC-seq replicas\n",
        "!wget -O ATAC-seq-1.bigBed \"https://www.encodeproject.org/files/ENCFF293GOJ/@@download/ENCFF293GOJ.bigBed\"\n",
        "!wget -O ATAC-seq-2.bigBed \"https://www.encodeproject.org/files/ENCFF155UDS/@@download/ENCFF155UDS.bigBed\"\n"
      ],
      "metadata": {
        "id": "9sH0T27MoWyY",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "a6d7a046-05cf-4189-9878-6669b970e126"
      },
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2024-11-23 13:25:22--  https://www.encodeproject.org/files/ENCFF629CUA/@@download/ENCFF629CUA.bigBed\n",
            "Resolving www.encodeproject.org (www.encodeproject.org)... 34.211.244.144\n",
            "Connecting to www.encodeproject.org (www.encodeproject.org)|34.211.244.144|:443... connected.\n",
            "HTTP request sent, awaiting response... 307 Temporary Redirect\n",
            "Location: https://encode-public.s3.amazonaws.com/2020/09/25/d88be941-9bb4-4a37-9718-75367e7af1e3/ENCFF629CUA.bigBed?response-content-disposition=attachment%3B%20filename%3DENCFF629CUA.bigBed&AWSAccessKeyId=ASIATGZNGCNXTKNWQ3PN&Signature=hvOnFfvHdiIpFzZPTRNhaDyYUnA%3D&x-amz-security-token=IQoJb3JpZ2luX2VjED0aCXVzLXdlc3QtMiJIMEYCIQDScuBbI8tnE9cu26KcU2kj4Y0mOSsx1Pak4RXtOag6bQIhAP2y3ei8rc5JqxNtsvuLpYDttKqXUTHbJBDZ4b%2Bbx8zVKrwFCNb%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEQABoMMjIwNzQ4NzE0ODYzIgyWvjQo1hCmTcQI4SAqkAXM9%2BPT5kGNbAwWujl9XyLiRwE0l4ZO2uwwqk2J9RVSRsqpkLJ6SgO18SpG0gBMTTTWzRwjOfPhzk%2BauH0C6vZpk4P4D5T9uGqVK8rXibkqeiiPXLrx4DqseOAlU0f%2F2YKJ%2Fy1UE%2Fm1m1LdKH9dwluf4GqNBGc7jCMzWKA1G8JZ3Cr4Cm%2BZD%2F%2BxvdHbmMFw2ltYYHd%2BVqJDdq%2FX4GPiFJS1kaS7%2FCjMF7lj%2FTOeIU0ETihRsnvnu9fWx2fMPOoIZoT3y%2FNoNsJssrxg0l3A0dCOIKTrp6nB%2Bzj1iLouswS4YpPiJw8XfQd%2FCNoAvzEQc%2BO1TeEMV%2B4p6291uugUWQN1IOZoIPGs%2Ffx1NtHzGaPZZmFIKtFx4coFywDrDcrtJoOJF0bA%2BgeRh6ME1onZIa0YOYriaN2U3PWyFbuxzuy3S2PuwSYh%2FHB%2F%2Buf%2F1Gp4pU%2FzKiT2Bg1LLmxD0JaHzBuuxPNHbA%2FzONEws3TPPH2cKu7nNi9Xyshzf5C%2Fa22nU4Vj9byYiLmnoaglaOqIiV0FAROVaLLw5tfTojQoFZ0nTFe9Y5sZ%2FNaaQ5m6wBUpV3DWvYdw0w8839I%2B1AaYyUhCYqoCrZVHC2gJP8ZRVSsgtIFe75WopDHs%2BuYgxG%2FYmF53Oa5%2BacI3iu0n%2B%2BOuyXhnriyrzFJBUSxXh9MTs0mwwvL0MSLUCa9Btmv2QWwlzrLCsL%2BTJ7ubf9otKgeG0cfAERV8GJS8jPZKT1SEz0eyLCq6r%2B3gdGYJbjqNrt2anKcyNSaEPFXLcEQh4BGmat7DcLIOaAxSRcgbBTBSISpZl4QDpESg7FLNg1lbjrVztnea%2Frt7FPvAkWSrdZgoTzL52WQx4cAJUhikPowGrg0cYzD1n4e6BjqwAfQVplb0qccE52BKH72rLykftsGjs%2B6nUbQYDoU6juZkpO67ddJ4tLDVp%2Fk2oEJOjOuQA7Nwm%2BZwObGQO22zHlbxf%2F5m%2B8LIfX4SQ%2B27qGm%2BDAs9Jko5ROxPdum5LwRYZDc5XxqkiYkt1Up8TwY4HBljWQXMpf%2BH3ngxvd37%2BTpJLMGuXtW%2BIiIVP7lZ4f90Fm7u4yt7VR7ffL0RutNdTSU5fkAotQSnsSnQ8iyBVdGP&Expires=1732497923 [following]\n",
            "--2024-11-23 13:25:23--  https://encode-public.s3.amazonaws.com/2020/09/25/d88be941-9bb4-4a37-9718-75367e7af1e3/ENCFF629CUA.bigBed?response-content-disposition=attachment%3B%20filename%3DENCFF629CUA.bigBed&AWSAccessKeyId=ASIATGZNGCNXTKNWQ3PN&Signature=hvOnFfvHdiIpFzZPTRNhaDyYUnA%3D&x-amz-security-token=IQoJb3JpZ2luX2VjED0aCXVzLXdlc3QtMiJIMEYCIQDScuBbI8tnE9cu26KcU2kj4Y0mOSsx1Pak4RXtOag6bQIhAP2y3ei8rc5JqxNtsvuLpYDttKqXUTHbJBDZ4b%2Bbx8zVKrwFCNb%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEQABoMMjIwNzQ4NzE0ODYzIgyWvjQo1hCmTcQI4SAqkAXM9%2BPT5kGNbAwWujl9XyLiRwE0l4ZO2uwwqk2J9RVSRsqpkLJ6SgO18SpG0gBMTTTWzRwjOfPhzk%2BauH0C6vZpk4P4D5T9uGqVK8rXibkqeiiPXLrx4DqseOAlU0f%2F2YKJ%2Fy1UE%2Fm1m1LdKH9dwluf4GqNBGc7jCMzWKA1G8JZ3Cr4Cm%2BZD%2F%2BxvdHbmMFw2ltYYHd%2BVqJDdq%2FX4GPiFJS1kaS7%2FCjMF7lj%2FTOeIU0ETihRsnvnu9fWx2fMPOoIZoT3y%2FNoNsJssrxg0l3A0dCOIKTrp6nB%2Bzj1iLouswS4YpPiJw8XfQd%2FCNoAvzEQc%2BO1TeEMV%2B4p6291uugUWQN1IOZoIPGs%2Ffx1NtHzGaPZZmFIKtFx4coFywDrDcrtJoOJF0bA%2BgeRh6ME1onZIa0YOYriaN2U3PWyFbuxzuy3S2PuwSYh%2FHB%2F%2Buf%2F1Gp4pU%2FzKiT2Bg1LLmxD0JaHzBuuxPNHbA%2FzONEws3TPPH2cKu7nNi9Xyshzf5C%2Fa22nU4Vj9byYiLmnoaglaOqIiV0FAROVaLLw5tfTojQoFZ0nTFe9Y5sZ%2FNaaQ5m6wBUpV3DWvYdw0w8839I%2B1AaYyUhCYqoCrZVHC2gJP8ZRVSsgtIFe75WopDHs%2BuYgxG%2FYmF53Oa5%2BacI3iu0n%2B%2BOuyXhnriyrzFJBUSxXh9MTs0mwwvL0MSLUCa9Btmv2QWwlzrLCsL%2BTJ7ubf9otKgeG0cfAERV8GJS8jPZKT1SEz0eyLCq6r%2B3gdGYJbjqNrt2anKcyNSaEPFXLcEQh4BGmat7DcLIOaAxSRcgbBTBSISpZl4QDpESg7FLNg1lbjrVztnea%2Frt7FPvAkWSrdZgoTzL52WQx4cAJUhikPowGrg0cYzD1n4e6BjqwAfQVplb0qccE52BKH72rLykftsGjs%2B6nUbQYDoU6juZkpO67ddJ4tLDVp%2Fk2oEJOjOuQA7Nwm%2BZwObGQO22zHlbxf%2F5m%2B8LIfX4SQ%2B27qGm%2BDAs9Jko5ROxPdum5LwRYZDc5XxqkiYkt1Up8TwY4HBljWQXMpf%2BH3ngxvd37%2BTpJLMGuXtW%2BIiIVP7lZ4f90Fm7u4yt7VR7ffL0RutNdTSU5fkAotQSnsSnQ8iyBVdGP&Expires=1732497923\n",
            "Resolving encode-public.s3.amazonaws.com (encode-public.s3.amazonaws.com)... 52.92.196.65, 52.92.160.145, 52.218.218.171, ...\n",
            "Connecting to encode-public.s3.amazonaws.com (encode-public.s3.amazonaws.com)|52.92.196.65|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 1451604 (1.4M) [binary/octet-stream]\n",
            "Saving to: ‘CTCF.bigBed’\n",
            "\n",
            "CTCF.bigBed         100%[===================>]   1.38M  3.32MB/s    in 0.4s    \n",
            "\n",
            "2024-11-23 13:25:24 (3.32 MB/s) - ‘CTCF.bigBed’ saved [1451604/1451604]\n",
            "\n",
            "--2024-11-23 13:25:24--  https://www.encodeproject.org/files/ENCFF671KDT/@@download/ENCFF671KDT.bigBed\n",
            "Resolving www.encodeproject.org (www.encodeproject.org)... 34.211.244.144\n",
            "Connecting to www.encodeproject.org (www.encodeproject.org)|34.211.244.144|:443... connected.\n",
            "HTTP request sent, awaiting response... 307 Temporary Redirect\n",
            "Location: https://encode-public.s3.amazonaws.com/2020/12/26/d1c35774-75cd-41c2-9c7f-8f58a8757690/ENCFF671KDT.bigBed?response-content-disposition=attachment%3B%20filename%3DENCFF671KDT.bigBed&AWSAccessKeyId=ASIATGZNGCNXTKNWQ3PN&Signature=mT1OI58A5g8PUL%2FSCdAGRRvdapQ%3D&x-amz-security-token=IQoJb3JpZ2luX2VjED0aCXVzLXdlc3QtMiJIMEYCIQDScuBbI8tnE9cu26KcU2kj4Y0mOSsx1Pak4RXtOag6bQIhAP2y3ei8rc5JqxNtsvuLpYDttKqXUTHbJBDZ4b%2Bbx8zVKrwFCNb%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEQABoMMjIwNzQ4NzE0ODYzIgyWvjQo1hCmTcQI4SAqkAXM9%2BPT5kGNbAwWujl9XyLiRwE0l4ZO2uwwqk2J9RVSRsqpkLJ6SgO18SpG0gBMTTTWzRwjOfPhzk%2BauH0C6vZpk4P4D5T9uGqVK8rXibkqeiiPXLrx4DqseOAlU0f%2F2YKJ%2Fy1UE%2Fm1m1LdKH9dwluf4GqNBGc7jCMzWKA1G8JZ3Cr4Cm%2BZD%2F%2BxvdHbmMFw2ltYYHd%2BVqJDdq%2FX4GPiFJS1kaS7%2FCjMF7lj%2FTOeIU0ETihRsnvnu9fWx2fMPOoIZoT3y%2FNoNsJssrxg0l3A0dCOIKTrp6nB%2Bzj1iLouswS4YpPiJw8XfQd%2FCNoAvzEQc%2BO1TeEMV%2B4p6291uugUWQN1IOZoIPGs%2Ffx1NtHzGaPZZmFIKtFx4coFywDrDcrtJoOJF0bA%2BgeRh6ME1onZIa0YOYriaN2U3PWyFbuxzuy3S2PuwSYh%2FHB%2F%2Buf%2F1Gp4pU%2FzKiT2Bg1LLmxD0JaHzBuuxPNHbA%2FzONEws3TPPH2cKu7nNi9Xyshzf5C%2Fa22nU4Vj9byYiLmnoaglaOqIiV0FAROVaLLw5tfTojQoFZ0nTFe9Y5sZ%2FNaaQ5m6wBUpV3DWvYdw0w8839I%2B1AaYyUhCYqoCrZVHC2gJP8ZRVSsgtIFe75WopDHs%2BuYgxG%2FYmF53Oa5%2BacI3iu0n%2B%2BOuyXhnriyrzFJBUSxXh9MTs0mwwvL0MSLUCa9Btmv2QWwlzrLCsL%2BTJ7ubf9otKgeG0cfAERV8GJS8jPZKT1SEz0eyLCq6r%2B3gdGYJbjqNrt2anKcyNSaEPFXLcEQh4BGmat7DcLIOaAxSRcgbBTBSISpZl4QDpESg7FLNg1lbjrVztnea%2Frt7FPvAkWSrdZgoTzL52WQx4cAJUhikPowGrg0cYzD1n4e6BjqwAfQVplb0qccE52BKH72rLykftsGjs%2B6nUbQYDoU6juZkpO67ddJ4tLDVp%2Fk2oEJOjOuQA7Nwm%2BZwObGQO22zHlbxf%2F5m%2B8LIfX4SQ%2B27qGm%2BDAs9Jko5ROxPdum5LwRYZDc5XxqkiYkt1Up8TwY4HBljWQXMpf%2BH3ngxvd37%2BTpJLMGuXtW%2BIiIVP7lZ4f90Fm7u4yt7VR7ffL0RutNdTSU5fkAotQSnsSnQ8iyBVdGP&Expires=1732497924 [following]\n",
            "--2024-11-23 13:25:24--  https://encode-public.s3.amazonaws.com/2020/12/26/d1c35774-75cd-41c2-9c7f-8f58a8757690/ENCFF671KDT.bigBed?response-content-disposition=attachment%3B%20filename%3DENCFF671KDT.bigBed&AWSAccessKeyId=ASIATGZNGCNXTKNWQ3PN&Signature=mT1OI58A5g8PUL%2FSCdAGRRvdapQ%3D&x-amz-security-token=IQoJb3JpZ2luX2VjED0aCXVzLXdlc3QtMiJIMEYCIQDScuBbI8tnE9cu26KcU2kj4Y0mOSsx1Pak4RXtOag6bQIhAP2y3ei8rc5JqxNtsvuLpYDttKqXUTHbJBDZ4b%2Bbx8zVKrwFCNb%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEQABoMMjIwNzQ4NzE0ODYzIgyWvjQo1hCmTcQI4SAqkAXM9%2BPT5kGNbAwWujl9XyLiRwE0l4ZO2uwwqk2J9RVSRsqpkLJ6SgO18SpG0gBMTTTWzRwjOfPhzk%2BauH0C6vZpk4P4D5T9uGqVK8rXibkqeiiPXLrx4DqseOAlU0f%2F2YKJ%2Fy1UE%2Fm1m1LdKH9dwluf4GqNBGc7jCMzWKA1G8JZ3Cr4Cm%2BZD%2F%2BxvdHbmMFw2ltYYHd%2BVqJDdq%2FX4GPiFJS1kaS7%2FCjMF7lj%2FTOeIU0ETihRsnvnu9fWx2fMPOoIZoT3y%2FNoNsJssrxg0l3A0dCOIKTrp6nB%2Bzj1iLouswS4YpPiJw8XfQd%2FCNoAvzEQc%2BO1TeEMV%2B4p6291uugUWQN1IOZoIPGs%2Ffx1NtHzGaPZZmFIKtFx4coFywDrDcrtJoOJF0bA%2BgeRh6ME1onZIa0YOYriaN2U3PWyFbuxzuy3S2PuwSYh%2FHB%2F%2Buf%2F1Gp4pU%2FzKiT2Bg1LLmxD0JaHzBuuxPNHbA%2FzONEws3TPPH2cKu7nNi9Xyshzf5C%2Fa22nU4Vj9byYiLmnoaglaOqIiV0FAROVaLLw5tfTojQoFZ0nTFe9Y5sZ%2FNaaQ5m6wBUpV3DWvYdw0w8839I%2B1AaYyUhCYqoCrZVHC2gJP8ZRVSsgtIFe75WopDHs%2BuYgxG%2FYmF53Oa5%2BacI3iu0n%2B%2BOuyXhnriyrzFJBUSxXh9MTs0mwwvL0MSLUCa9Btmv2QWwlzrLCsL%2BTJ7ubf9otKgeG0cfAERV8GJS8jPZKT1SEz0eyLCq6r%2B3gdGYJbjqNrt2anKcyNSaEPFXLcEQh4BGmat7DcLIOaAxSRcgbBTBSISpZl4QDpESg7FLNg1lbjrVztnea%2Frt7FPvAkWSrdZgoTzL52WQx4cAJUhikPowGrg0cYzD1n4e6BjqwAfQVplb0qccE52BKH72rLykftsGjs%2B6nUbQYDoU6juZkpO67ddJ4tLDVp%2Fk2oEJOjOuQA7Nwm%2BZwObGQO22zHlbxf%2F5m%2B8LIfX4SQ%2B27qGm%2BDAs9Jko5ROxPdum5LwRYZDc5XxqkiYkt1Up8TwY4HBljWQXMpf%2BH3ngxvd37%2BTpJLMGuXtW%2BIiIVP7lZ4f90Fm7u4yt7VR7ffL0RutNdTSU5fkAotQSnsSnQ8iyBVdGP&Expires=1732497924\n",
            "Resolving encode-public.s3.amazonaws.com (encode-public.s3.amazonaws.com)... 52.92.196.65, 52.92.160.145, 52.218.218.171, ...\n",
            "Connecting to encode-public.s3.amazonaws.com (encode-public.s3.amazonaws.com)|52.92.196.65|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 539538 (527K) [binary/octet-stream]\n",
            "Saving to: ‘EP300.bigBed’\n",
            "\n",
            "EP300.bigBed        100%[===================>] 526.89K  1.43MB/s    in 0.4s    \n",
            "\n",
            "2024-11-23 13:25:25 (1.43 MB/s) - ‘EP300.bigBed’ saved [539538/539538]\n",
            "\n",
            "--2024-11-23 13:25:25--  https://www.encodeproject.org/files/ENCFF927NKT/@@download/ENCFF927NKT.bigBed\n",
            "Resolving www.encodeproject.org (www.encodeproject.org)... 34.211.244.144\n",
            "Connecting to www.encodeproject.org (www.encodeproject.org)|34.211.244.144|:443... connected.\n",
            "HTTP request sent, awaiting response... 307 Temporary Redirect\n",
            "Location: https://encode-public.s3.amazonaws.com/2020/11/29/470b8294-421e-4e73-8d2f-02f21c9cdc00/ENCFF927NKT.bigBed?response-content-disposition=attachment%3B%20filename%3DENCFF927NKT.bigBed&AWSAccessKeyId=ASIATGZNGCNXTKNWQ3PN&Signature=aKqo4YIsob6lQ8H2kcCtWZi3ndA%3D&x-amz-security-token=IQoJb3JpZ2luX2VjED0aCXVzLXdlc3QtMiJIMEYCIQDScuBbI8tnE9cu26KcU2kj4Y0mOSsx1Pak4RXtOag6bQIhAP2y3ei8rc5JqxNtsvuLpYDttKqXUTHbJBDZ4b%2Bbx8zVKrwFCNb%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEQABoMMjIwNzQ4NzE0ODYzIgyWvjQo1hCmTcQI4SAqkAXM9%2BPT5kGNbAwWujl9XyLiRwE0l4ZO2uwwqk2J9RVSRsqpkLJ6SgO18SpG0gBMTTTWzRwjOfPhzk%2BauH0C6vZpk4P4D5T9uGqVK8rXibkqeiiPXLrx4DqseOAlU0f%2F2YKJ%2Fy1UE%2Fm1m1LdKH9dwluf4GqNBGc7jCMzWKA1G8JZ3Cr4Cm%2BZD%2F%2BxvdHbmMFw2ltYYHd%2BVqJDdq%2FX4GPiFJS1kaS7%2FCjMF7lj%2FTOeIU0ETihRsnvnu9fWx2fMPOoIZoT3y%2FNoNsJssrxg0l3A0dCOIKTrp6nB%2Bzj1iLouswS4YpPiJw8XfQd%2FCNoAvzEQc%2BO1TeEMV%2B4p6291uugUWQN1IOZoIPGs%2Ffx1NtHzGaPZZmFIKtFx4coFywDrDcrtJoOJF0bA%2BgeRh6ME1onZIa0YOYriaN2U3PWyFbuxzuy3S2PuwSYh%2FHB%2F%2Buf%2F1Gp4pU%2FzKiT2Bg1LLmxD0JaHzBuuxPNHbA%2FzONEws3TPPH2cKu7nNi9Xyshzf5C%2Fa22nU4Vj9byYiLmnoaglaOqIiV0FAROVaLLw5tfTojQoFZ0nTFe9Y5sZ%2FNaaQ5m6wBUpV3DWvYdw0w8839I%2B1AaYyUhCYqoCrZVHC2gJP8ZRVSsgtIFe75WopDHs%2BuYgxG%2FYmF53Oa5%2BacI3iu0n%2B%2BOuyXhnriyrzFJBUSxXh9MTs0mwwvL0MSLUCa9Btmv2QWwlzrLCsL%2BTJ7ubf9otKgeG0cfAERV8GJS8jPZKT1SEz0eyLCq6r%2B3gdGYJbjqNrt2anKcyNSaEPFXLcEQh4BGmat7DcLIOaAxSRcgbBTBSISpZl4QDpESg7FLNg1lbjrVztnea%2Frt7FPvAkWSrdZgoTzL52WQx4cAJUhikPowGrg0cYzD1n4e6BjqwAfQVplb0qccE52BKH72rLykftsGjs%2B6nUbQYDoU6juZkpO67ddJ4tLDVp%2Fk2oEJOjOuQA7Nwm%2BZwObGQO22zHlbxf%2F5m%2B8LIfX4SQ%2B27qGm%2BDAs9Jko5ROxPdum5LwRYZDc5XxqkiYkt1Up8TwY4HBljWQXMpf%2BH3ngxvd37%2BTpJLMGuXtW%2BIiIVP7lZ4f90Fm7u4yt7VR7ffL0RutNdTSU5fkAotQSnsSnQ8iyBVdGP&Expires=1732497925 [following]\n",
            "--2024-11-23 13:25:25--  https://encode-public.s3.amazonaws.com/2020/11/29/470b8294-421e-4e73-8d2f-02f21c9cdc00/ENCFF927NKT.bigBed?response-content-disposition=attachment%3B%20filename%3DENCFF927NKT.bigBed&AWSAccessKeyId=ASIATGZNGCNXTKNWQ3PN&Signature=aKqo4YIsob6lQ8H2kcCtWZi3ndA%3D&x-amz-security-token=IQoJb3JpZ2luX2VjED0aCXVzLXdlc3QtMiJIMEYCIQDScuBbI8tnE9cu26KcU2kj4Y0mOSsx1Pak4RXtOag6bQIhAP2y3ei8rc5JqxNtsvuLpYDttKqXUTHbJBDZ4b%2Bbx8zVKrwFCNb%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEQABoMMjIwNzQ4NzE0ODYzIgyWvjQo1hCmTcQI4SAqkAXM9%2BPT5kGNbAwWujl9XyLiRwE0l4ZO2uwwqk2J9RVSRsqpkLJ6SgO18SpG0gBMTTTWzRwjOfPhzk%2BauH0C6vZpk4P4D5T9uGqVK8rXibkqeiiPXLrx4DqseOAlU0f%2F2YKJ%2Fy1UE%2Fm1m1LdKH9dwluf4GqNBGc7jCMzWKA1G8JZ3Cr4Cm%2BZD%2F%2BxvdHbmMFw2ltYYHd%2BVqJDdq%2FX4GPiFJS1kaS7%2FCjMF7lj%2FTOeIU0ETihRsnvnu9fWx2fMPOoIZoT3y%2FNoNsJssrxg0l3A0dCOIKTrp6nB%2Bzj1iLouswS4YpPiJw8XfQd%2FCNoAvzEQc%2BO1TeEMV%2B4p6291uugUWQN1IOZoIPGs%2Ffx1NtHzGaPZZmFIKtFx4coFywDrDcrtJoOJF0bA%2BgeRh6ME1onZIa0YOYriaN2U3PWyFbuxzuy3S2PuwSYh%2FHB%2F%2Buf%2F1Gp4pU%2FzKiT2Bg1LLmxD0JaHzBuuxPNHbA%2FzONEws3TPPH2cKu7nNi9Xyshzf5C%2Fa22nU4Vj9byYiLmnoaglaOqIiV0FAROVaLLw5tfTojQoFZ0nTFe9Y5sZ%2FNaaQ5m6wBUpV3DWvYdw0w8839I%2B1AaYyUhCYqoCrZVHC2gJP8ZRVSsgtIFe75WopDHs%2BuYgxG%2FYmF53Oa5%2BacI3iu0n%2B%2BOuyXhnriyrzFJBUSxXh9MTs0mwwvL0MSLUCa9Btmv2QWwlzrLCsL%2BTJ7ubf9otKgeG0cfAERV8GJS8jPZKT1SEz0eyLCq6r%2B3gdGYJbjqNrt2anKcyNSaEPFXLcEQh4BGmat7DcLIOaAxSRcgbBTBSISpZl4QDpESg7FLNg1lbjrVztnea%2Frt7FPvAkWSrdZgoTzL52WQx4cAJUhikPowGrg0cYzD1n4e6BjqwAfQVplb0qccE52BKH72rLykftsGjs%2B6nUbQYDoU6juZkpO67ddJ4tLDVp%2Fk2oEJOjOuQA7Nwm%2BZwObGQO22zHlbxf%2F5m%2B8LIfX4SQ%2B27qGm%2BDAs9Jko5ROxPdum5LwRYZDc5XxqkiYkt1Up8TwY4HBljWQXMpf%2BH3ngxvd37%2BTpJLMGuXtW%2BIiIVP7lZ4f90Fm7u4yt7VR7ffL0RutNdTSU5fkAotQSnsSnQ8iyBVdGP&Expires=1732497925\n",
            "Resolving encode-public.s3.amazonaws.com (encode-public.s3.amazonaws.com)... 52.92.196.65, 52.92.160.145, 52.218.218.171, ...\n",
            "Connecting to encode-public.s3.amazonaws.com (encode-public.s3.amazonaws.com)|52.92.196.65|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 637768 (623K) [binary/octet-stream]\n",
            "Saving to: ‘POLR2A.bigBed’\n",
            "\n",
            "POLR2A.bigBed       100%[===================>] 622.82K  1.72MB/s    in 0.4s    \n",
            "\n",
            "2024-11-23 13:25:26 (1.72 MB/s) - ‘POLR2A.bigBed’ saved [637768/637768]\n",
            "\n",
            "--2024-11-23 13:25:26--  https://www.encodeproject.org/files/ENCFF293GOJ/@@download/ENCFF293GOJ.bigBed\n",
            "Resolving www.encodeproject.org (www.encodeproject.org)... 34.211.244.144\n",
            "Connecting to www.encodeproject.org (www.encodeproject.org)|34.211.244.144|:443... connected.\n",
            "HTTP request sent, awaiting response... 307 Temporary Redirect\n",
            "Location: https://encode-public.s3.amazonaws.com/2021/02/24/b3f4b66f-a080-4085-a64b-6f862a2cf618/ENCFF293GOJ.bigBed?response-content-disposition=attachment%3B%20filename%3DENCFF293GOJ.bigBed&AWSAccessKeyId=ASIATGZNGCNXTKNWQ3PN&Signature=r8nmuzjq0q4Wj3g4rZ%2Fr%2Fr5dDi4%3D&x-amz-security-token=IQoJb3JpZ2luX2VjED0aCXVzLXdlc3QtMiJIMEYCIQDScuBbI8tnE9cu26KcU2kj4Y0mOSsx1Pak4RXtOag6bQIhAP2y3ei8rc5JqxNtsvuLpYDttKqXUTHbJBDZ4b%2Bbx8zVKrwFCNb%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEQABoMMjIwNzQ4NzE0ODYzIgyWvjQo1hCmTcQI4SAqkAXM9%2BPT5kGNbAwWujl9XyLiRwE0l4ZO2uwwqk2J9RVSRsqpkLJ6SgO18SpG0gBMTTTWzRwjOfPhzk%2BauH0C6vZpk4P4D5T9uGqVK8rXibkqeiiPXLrx4DqseOAlU0f%2F2YKJ%2Fy1UE%2Fm1m1LdKH9dwluf4GqNBGc7jCMzWKA1G8JZ3Cr4Cm%2BZD%2F%2BxvdHbmMFw2ltYYHd%2BVqJDdq%2FX4GPiFJS1kaS7%2FCjMF7lj%2FTOeIU0ETihRsnvnu9fWx2fMPOoIZoT3y%2FNoNsJssrxg0l3A0dCOIKTrp6nB%2Bzj1iLouswS4YpPiJw8XfQd%2FCNoAvzEQc%2BO1TeEMV%2B4p6291uugUWQN1IOZoIPGs%2Ffx1NtHzGaPZZmFIKtFx4coFywDrDcrtJoOJF0bA%2BgeRh6ME1onZIa0YOYriaN2U3PWyFbuxzuy3S2PuwSYh%2FHB%2F%2Buf%2F1Gp4pU%2FzKiT2Bg1LLmxD0JaHzBuuxPNHbA%2FzONEws3TPPH2cKu7nNi9Xyshzf5C%2Fa22nU4Vj9byYiLmnoaglaOqIiV0FAROVaLLw5tfTojQoFZ0nTFe9Y5sZ%2FNaaQ5m6wBUpV3DWvYdw0w8839I%2B1AaYyUhCYqoCrZVHC2gJP8ZRVSsgtIFe75WopDHs%2BuYgxG%2FYmF53Oa5%2BacI3iu0n%2B%2BOuyXhnriyrzFJBUSxXh9MTs0mwwvL0MSLUCa9Btmv2QWwlzrLCsL%2BTJ7ubf9otKgeG0cfAERV8GJS8jPZKT1SEz0eyLCq6r%2B3gdGYJbjqNrt2anKcyNSaEPFXLcEQh4BGmat7DcLIOaAxSRcgbBTBSISpZl4QDpESg7FLNg1lbjrVztnea%2Frt7FPvAkWSrdZgoTzL52WQx4cAJUhikPowGrg0cYzD1n4e6BjqwAfQVplb0qccE52BKH72rLykftsGjs%2B6nUbQYDoU6juZkpO67ddJ4tLDVp%2Fk2oEJOjOuQA7Nwm%2BZwObGQO22zHlbxf%2F5m%2B8LIfX4SQ%2B27qGm%2BDAs9Jko5ROxPdum5LwRYZDc5XxqkiYkt1Up8TwY4HBljWQXMpf%2BH3ngxvd37%2BTpJLMGuXtW%2BIiIVP7lZ4f90Fm7u4yt7VR7ffL0RutNdTSU5fkAotQSnsSnQ8iyBVdGP&Expires=1732497926 [following]\n",
            "--2024-11-23 13:25:26--  https://encode-public.s3.amazonaws.com/2021/02/24/b3f4b66f-a080-4085-a64b-6f862a2cf618/ENCFF293GOJ.bigBed?response-content-disposition=attachment%3B%20filename%3DENCFF293GOJ.bigBed&AWSAccessKeyId=ASIATGZNGCNXTKNWQ3PN&Signature=r8nmuzjq0q4Wj3g4rZ%2Fr%2Fr5dDi4%3D&x-amz-security-token=IQoJb3JpZ2luX2VjED0aCXVzLXdlc3QtMiJIMEYCIQDScuBbI8tnE9cu26KcU2kj4Y0mOSsx1Pak4RXtOag6bQIhAP2y3ei8rc5JqxNtsvuLpYDttKqXUTHbJBDZ4b%2Bbx8zVKrwFCNb%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEQABoMMjIwNzQ4NzE0ODYzIgyWvjQo1hCmTcQI4SAqkAXM9%2BPT5kGNbAwWujl9XyLiRwE0l4ZO2uwwqk2J9RVSRsqpkLJ6SgO18SpG0gBMTTTWzRwjOfPhzk%2BauH0C6vZpk4P4D5T9uGqVK8rXibkqeiiPXLrx4DqseOAlU0f%2F2YKJ%2Fy1UE%2Fm1m1LdKH9dwluf4GqNBGc7jCMzWKA1G8JZ3Cr4Cm%2BZD%2F%2BxvdHbmMFw2ltYYHd%2BVqJDdq%2FX4GPiFJS1kaS7%2FCjMF7lj%2FTOeIU0ETihRsnvnu9fWx2fMPOoIZoT3y%2FNoNsJssrxg0l3A0dCOIKTrp6nB%2Bzj1iLouswS4YpPiJw8XfQd%2FCNoAvzEQc%2BO1TeEMV%2B4p6291uugUWQN1IOZoIPGs%2Ffx1NtHzGaPZZmFIKtFx4coFywDrDcrtJoOJF0bA%2BgeRh6ME1onZIa0YOYriaN2U3PWyFbuxzuy3S2PuwSYh%2FHB%2F%2Buf%2F1Gp4pU%2FzKiT2Bg1LLmxD0JaHzBuuxPNHbA%2FzONEws3TPPH2cKu7nNi9Xyshzf5C%2Fa22nU4Vj9byYiLmnoaglaOqIiV0FAROVaLLw5tfTojQoFZ0nTFe9Y5sZ%2FNaaQ5m6wBUpV3DWvYdw0w8839I%2B1AaYyUhCYqoCrZVHC2gJP8ZRVSsgtIFe75WopDHs%2BuYgxG%2FYmF53Oa5%2BacI3iu0n%2B%2BOuyXhnriyrzFJBUSxXh9MTs0mwwvL0MSLUCa9Btmv2QWwlzrLCsL%2BTJ7ubf9otKgeG0cfAERV8GJS8jPZKT1SEz0eyLCq6r%2B3gdGYJbjqNrt2anKcyNSaEPFXLcEQh4BGmat7DcLIOaAxSRcgbBTBSISpZl4QDpESg7FLNg1lbjrVztnea%2Frt7FPvAkWSrdZgoTzL52WQx4cAJUhikPowGrg0cYzD1n4e6BjqwAfQVplb0qccE52BKH72rLykftsGjs%2B6nUbQYDoU6juZkpO67ddJ4tLDVp%2Fk2oEJOjOuQA7Nwm%2BZwObGQO22zHlbxf%2F5m%2B8LIfX4SQ%2B27qGm%2BDAs9Jko5ROxPdum5LwRYZDc5XxqkiYkt1Up8TwY4HBljWQXMpf%2BH3ngxvd37%2BTpJLMGuXtW%2BIiIVP7lZ4f90Fm7u4yt7VR7ffL0RutNdTSU5fkAotQSnsSnQ8iyBVdGP&Expires=1732497926\n",
            "Resolving encode-public.s3.amazonaws.com (encode-public.s3.amazonaws.com)... 52.92.196.65, 52.92.160.145, 52.218.218.171, ...\n",
            "Connecting to encode-public.s3.amazonaws.com (encode-public.s3.amazonaws.com)|52.92.196.65|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 4339188 (4.1M) [binary/octet-stream]\n",
            "Saving to: ‘ATAC-seq-1.bigBed’\n",
            "\n",
            "ATAC-seq-1.bigBed   100%[===================>]   4.14M  7.30MB/s    in 0.6s    \n",
            "\n",
            "2024-11-23 13:25:27 (7.30 MB/s) - ‘ATAC-seq-1.bigBed’ saved [4339188/4339188]\n",
            "\n",
            "--2024-11-23 13:25:27--  https://www.encodeproject.org/files/ENCFF155UDS/@@download/ENCFF155UDS.bigBed\n",
            "Resolving www.encodeproject.org (www.encodeproject.org)... 34.211.244.144\n",
            "Connecting to www.encodeproject.org (www.encodeproject.org)|34.211.244.144|:443... connected.\n",
            "HTTP request sent, awaiting response... 307 Temporary Redirect\n",
            "Location: https://encode-public.s3.amazonaws.com/2021/02/24/56626323-db4c-40d8-a556-521eabe352bc/ENCFF155UDS.bigBed?response-content-disposition=attachment%3B%20filename%3DENCFF155UDS.bigBed&AWSAccessKeyId=ASIATGZNGCNXTKNWQ3PN&Signature=15BqbqR0IRjkeFJKhWKVBtJKnPQ%3D&x-amz-security-token=IQoJb3JpZ2luX2VjED0aCXVzLXdlc3QtMiJIMEYCIQDScuBbI8tnE9cu26KcU2kj4Y0mOSsx1Pak4RXtOag6bQIhAP2y3ei8rc5JqxNtsvuLpYDttKqXUTHbJBDZ4b%2Bbx8zVKrwFCNb%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEQABoMMjIwNzQ4NzE0ODYzIgyWvjQo1hCmTcQI4SAqkAXM9%2BPT5kGNbAwWujl9XyLiRwE0l4ZO2uwwqk2J9RVSRsqpkLJ6SgO18SpG0gBMTTTWzRwjOfPhzk%2BauH0C6vZpk4P4D5T9uGqVK8rXibkqeiiPXLrx4DqseOAlU0f%2F2YKJ%2Fy1UE%2Fm1m1LdKH9dwluf4GqNBGc7jCMzWKA1G8JZ3Cr4Cm%2BZD%2F%2BxvdHbmMFw2ltYYHd%2BVqJDdq%2FX4GPiFJS1kaS7%2FCjMF7lj%2FTOeIU0ETihRsnvnu9fWx2fMPOoIZoT3y%2FNoNsJssrxg0l3A0dCOIKTrp6nB%2Bzj1iLouswS4YpPiJw8XfQd%2FCNoAvzEQc%2BO1TeEMV%2B4p6291uugUWQN1IOZoIPGs%2Ffx1NtHzGaPZZmFIKtFx4coFywDrDcrtJoOJF0bA%2BgeRh6ME1onZIa0YOYriaN2U3PWyFbuxzuy3S2PuwSYh%2FHB%2F%2Buf%2F1Gp4pU%2FzKiT2Bg1LLmxD0JaHzBuuxPNHbA%2FzONEws3TPPH2cKu7nNi9Xyshzf5C%2Fa22nU4Vj9byYiLmnoaglaOqIiV0FAROVaLLw5tfTojQoFZ0nTFe9Y5sZ%2FNaaQ5m6wBUpV3DWvYdw0w8839I%2B1AaYyUhCYqoCrZVHC2gJP8ZRVSsgtIFe75WopDHs%2BuYgxG%2FYmF53Oa5%2BacI3iu0n%2B%2BOuyXhnriyrzFJBUSxXh9MTs0mwwvL0MSLUCa9Btmv2QWwlzrLCsL%2BTJ7ubf9otKgeG0cfAERV8GJS8jPZKT1SEz0eyLCq6r%2B3gdGYJbjqNrt2anKcyNSaEPFXLcEQh4BGmat7DcLIOaAxSRcgbBTBSISpZl4QDpESg7FLNg1lbjrVztnea%2Frt7FPvAkWSrdZgoTzL52WQx4cAJUhikPowGrg0cYzD1n4e6BjqwAfQVplb0qccE52BKH72rLykftsGjs%2B6nUbQYDoU6juZkpO67ddJ4tLDVp%2Fk2oEJOjOuQA7Nwm%2BZwObGQO22zHlbxf%2F5m%2B8LIfX4SQ%2B27qGm%2BDAs9Jko5ROxPdum5LwRYZDc5XxqkiYkt1Up8TwY4HBljWQXMpf%2BH3ngxvd37%2BTpJLMGuXtW%2BIiIVP7lZ4f90Fm7u4yt7VR7ffL0RutNdTSU5fkAotQSnsSnQ8iyBVdGP&Expires=1732497927 [following]\n",
            "--2024-11-23 13:25:27--  https://encode-public.s3.amazonaws.com/2021/02/24/56626323-db4c-40d8-a556-521eabe352bc/ENCFF155UDS.bigBed?response-content-disposition=attachment%3B%20filename%3DENCFF155UDS.bigBed&AWSAccessKeyId=ASIATGZNGCNXTKNWQ3PN&Signature=15BqbqR0IRjkeFJKhWKVBtJKnPQ%3D&x-amz-security-token=IQoJb3JpZ2luX2VjED0aCXVzLXdlc3QtMiJIMEYCIQDScuBbI8tnE9cu26KcU2kj4Y0mOSsx1Pak4RXtOag6bQIhAP2y3ei8rc5JqxNtsvuLpYDttKqXUTHbJBDZ4b%2Bbx8zVKrwFCNb%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEQABoMMjIwNzQ4NzE0ODYzIgyWvjQo1hCmTcQI4SAqkAXM9%2BPT5kGNbAwWujl9XyLiRwE0l4ZO2uwwqk2J9RVSRsqpkLJ6SgO18SpG0gBMTTTWzRwjOfPhzk%2BauH0C6vZpk4P4D5T9uGqVK8rXibkqeiiPXLrx4DqseOAlU0f%2F2YKJ%2Fy1UE%2Fm1m1LdKH9dwluf4GqNBGc7jCMzWKA1G8JZ3Cr4Cm%2BZD%2F%2BxvdHbmMFw2ltYYHd%2BVqJDdq%2FX4GPiFJS1kaS7%2FCjMF7lj%2FTOeIU0ETihRsnvnu9fWx2fMPOoIZoT3y%2FNoNsJssrxg0l3A0dCOIKTrp6nB%2Bzj1iLouswS4YpPiJw8XfQd%2FCNoAvzEQc%2BO1TeEMV%2B4p6291uugUWQN1IOZoIPGs%2Ffx1NtHzGaPZZmFIKtFx4coFywDrDcrtJoOJF0bA%2BgeRh6ME1onZIa0YOYriaN2U3PWyFbuxzuy3S2PuwSYh%2FHB%2F%2Buf%2F1Gp4pU%2FzKiT2Bg1LLmxD0JaHzBuuxPNHbA%2FzONEws3TPPH2cKu7nNi9Xyshzf5C%2Fa22nU4Vj9byYiLmnoaglaOqIiV0FAROVaLLw5tfTojQoFZ0nTFe9Y5sZ%2FNaaQ5m6wBUpV3DWvYdw0w8839I%2B1AaYyUhCYqoCrZVHC2gJP8ZRVSsgtIFe75WopDHs%2BuYgxG%2FYmF53Oa5%2BacI3iu0n%2B%2BOuyXhnriyrzFJBUSxXh9MTs0mwwvL0MSLUCa9Btmv2QWwlzrLCsL%2BTJ7ubf9otKgeG0cfAERV8GJS8jPZKT1SEz0eyLCq6r%2B3gdGYJbjqNrt2anKcyNSaEPFXLcEQh4BGmat7DcLIOaAxSRcgbBTBSISpZl4QDpESg7FLNg1lbjrVztnea%2Frt7FPvAkWSrdZgoTzL52WQx4cAJUhikPowGrg0cYzD1n4e6BjqwAfQVplb0qccE52BKH72rLykftsGjs%2B6nUbQYDoU6juZkpO67ddJ4tLDVp%2Fk2oEJOjOuQA7Nwm%2BZwObGQO22zHlbxf%2F5m%2B8LIfX4SQ%2B27qGm%2BDAs9Jko5ROxPdum5LwRYZDc5XxqkiYkt1Up8TwY4HBljWQXMpf%2BH3ngxvd37%2BTpJLMGuXtW%2BIiIVP7lZ4f90Fm7u4yt7VR7ffL0RutNdTSU5fkAotQSnsSnQ8iyBVdGP&Expires=1732497927\n",
            "Resolving encode-public.s3.amazonaws.com (encode-public.s3.amazonaws.com)... 52.92.196.65, 52.92.160.145, 52.218.218.171, ...\n",
            "Connecting to encode-public.s3.amazonaws.com (encode-public.s3.amazonaws.com)|52.92.196.65|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 3899839 (3.7M) [binary/octet-stream]\n",
            "Saving to: ‘ATAC-seq-2.bigBed’\n",
            "\n",
            "ATAC-seq-2.bigBed   100%[===================>]   3.72M  7.40MB/s    in 0.5s    \n",
            "\n",
            "2024-11-23 13:25:28 (7.40 MB/s) - ‘ATAC-seq-2.bigBed’ saved [3899839/3899839]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Download the tool from the UCSC repository\n",
        "!wget http://hgdownload.cse.ucsc.edu/admin/exe/linux.x86_64/bigBedToBed\n",
        "# Add flag to allow the execution\n",
        "!chmod a+x bigBedToBed\n",
        "# Convert files one-by-one\n",
        "for file in \"CTCF\", \"EP300\", \"POLR2A\", \"ATAC-seq-1\", \"ATAC-seq-2\":\n",
        "  !./bigBedToBed \"{file}.bigBed\" \"{file}.bed\""
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "YFvAA-Z_QSJJ",
        "outputId": "69580d6b-8473-4edd-98d2-82b5fec44d18"
      },
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2024-11-23 13:25:29--  http://hgdownload.cse.ucsc.edu/admin/exe/linux.x86_64/bigBedToBed\n",
            "Resolving hgdownload.cse.ucsc.edu (hgdownload.cse.ucsc.edu)... 128.114.119.163\n",
            "Connecting to hgdownload.cse.ucsc.edu (hgdownload.cse.ucsc.edu)|128.114.119.163|:80... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 9492368 (9.1M)\n",
            "Saving to: ‘bigBedToBed’\n",
            "\n",
            "bigBedToBed         100%[===================>]   9.05M  13.0MB/s    in 0.7s    \n",
            "\n",
            "2024-11-23 13:25:30 (13.0 MB/s) - ‘bigBedToBed’ saved [9492368/9492368]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!head \"CTCF.bed\""
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "jsv6Mbp0QUMj",
        "outputId": "29edf7f7-be31-48d6-dfa3-a890a4c217eb"
      },
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "chr1\t16129\t16365\t.\t626\t.\t13.85036\t-1.00000\t3.77946\t118\n",
            "chr1\t267889\t268125\t.\t1000\t.\t41.92171\t-1.00000\t4.96567\t118\n",
            "chr1\t586084\t586320\t.\t1000\t.\t26.51044\t-1.00000\t4.96567\t118\n",
            "chr1\t778793\t778916\t.\t1000\t.\t68.56819\t-1.00000\t4.96567\t114\n",
            "chr1\t850462\t850698\t.\t699\t.\t17.31760\t-1.00000\t4.96567\t118\n",
            "chr1\t869856\t869987\t.\t1000\t.\t103.19001\t-1.00000\t4.96567\t57\n",
            "chr1\t904692\t904861\t.\t1000\t.\t127.98478\t-1.00000\t4.96567\t79\n",
            "chr1\t912903\t913139\t.\t1000\t.\t20.42124\t-1.00000\t4.96567\t118\n",
            "chr1\t921115\t921351\t.\t1000\t.\t44.15032\t-1.00000\t4.96567\t118\n",
            "chr1\t931656\t931892\t.\t575\t.\t10.13823\t-1.00000\t2.89810\t118\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Install bedtools using apt (Linux packet manager)\n",
        "!apt install -y bedtools\n",
        "# Install python wrapper using pip (Python packet manager)\n",
        "!pip3 install pybedtools"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "hhNmWqvyXaq9",
        "outputId": "46673c71-fa06-485a-88f3-ec89408b6368"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "The following NEW packages will be installed:\n",
            "  bedtools\n",
            "0 upgraded, 1 newly installed, 0 to remove and 49 not upgraded.\n",
            "Need to get 563 kB of archives.\n",
            "After this operation, 1,548 kB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu jammy-updates/universe amd64 bedtools amd64 2.30.0+dfsg-2ubuntu0.1 [563 kB]\n",
            "Fetched 563 kB in 1s (748 kB/s)\n",
            "Selecting previously unselected package bedtools.\n",
            "(Reading database ... 123630 files and directories currently installed.)\n",
            "Preparing to unpack .../bedtools_2.30.0+dfsg-2ubuntu0.1_amd64.deb ...\n",
            "Unpacking bedtools (2.30.0+dfsg-2ubuntu0.1) ...\n",
            "Setting up bedtools (2.30.0+dfsg-2ubuntu0.1) ...\n",
            "Collecting pybedtools\n",
            "  Downloading pybedtools-0.10.0.tar.gz (12.6 MB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m12.6/12.6 MB\u001b[0m \u001b[31m84.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "Collecting pysam (from pybedtools)\n",
            "  Downloading pysam-0.22.1-cp310-cp310-manylinux_2_28_x86_64.whl.metadata (1.5 kB)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.10/dist-packages (from pybedtools) (1.26.4)\n",
            "Downloading pysam-0.22.1-cp310-cp310-manylinux_2_28_x86_64.whl (22.0 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m22.0/22.0 MB\u001b[0m \u001b[31m66.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hBuilding wheels for collected packages: pybedtools\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "from pybedtools import BedTool\n",
        "\n",
        "# .sort() is required to effectively subtract/intersect regions later\n",
        "ctcf = BedTool(\"CTCF.bed\").sort()\n",
        "ep300 = BedTool(\"EP300.bed\").sort()\n",
        "polr2a = BedTool(\"POLR2A.bed\").sort()\n",
        "\n",
        "atac1 = BedTool(\"ATAC-seq-1.bed\").sort()\n",
        "atac2 = BedTool(\"ATAC-seq-2.bed\").sort()\n"
      ],
      "metadata": {
        "id": "4EFl-NopVhaG"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# The final set of ATAC-seq peaks is the intersection of two replicas\n",
        "atac = atac1.intersect(atac2).sort()\n",
        "\n",
        "# Highlighting regions not reproduced between ATAC-seq replicas\n",
        "atac1_only = atac1.subtract(atac2).sort()\n",
        "atac2_only = atac2.subtract(atac1).sort()\n",
        "atac_not_replicated = atac1_only.cat(atac2_only).sort()\n",
        "\n",
        "# Removing weak ATAC-seq peaks from each ChIP-seq file\n",
        "ctcf = ctcf.subtract(atac_not_replicated).sort()\n",
        "ep300 = ep300.subtract(atac_not_replicated).sort()\n",
        "polr2a = polr2a.subtract(atac_not_replicated).sort()\n",
        "\n",
        "# Checking that the intersection regions do not contain non-reproduced regions\n",
        "assert atac.intersect(atac_not_replicated).total_coverage() == 0\n",
        "\n",
        "# Remove overlapping regions between TFs\n",
        "ctcf_unique = ctcf.subtract(ep300).subtract(polr2a).sort()\n",
        "ep300_unique = ep300.subtract(ctcf).subtract(polr2a).sort()\n",
        "polr2a_unique = polr2a.subtract(ctcf).subtract(ep300).sort()\n",
        "\n",
        "# Sanity check: Ensure no overlaps between TFs\n",
        "assert ctcf_unique.intersect(ep300_unique).total_coverage() == 0\n",
        "assert ep300_unique.intersect(polr2a_unique).total_coverage() == 0\n",
        "assert polr2a_unique.intersect(ctcf_unique).total_coverage() == 0"
      ],
      "metadata": {
        "id": "rZMv33j-Vh5B"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Now we can formalize 2 classes:\n",
        "* **Background (class 0):** open chromatin regions not associated with CTCF (ATAC peaks without CTCF peaks).\n",
        "* **Foreground for CTCF(class 1):** CTCF bound open chromatin (ATAC peaks with CTCF peaks).\n",
        "* **Foreground for EP300(class 2):** EP300 bound open chromatin (ATAC peaks with EP300 peaks).\n",
        "* **Foreground for POLR2A(class 3):** POLR2A bound open chromatin (ATAC peaks with POLR2A peaks)."
      ],
      "metadata": {
        "id": "p_gEN8haamXg"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "fg_ctcf = ctcf_unique.intersect(atac, wa=True, u=True).sort()\n",
        "fg_ep300 = ep300_unique.intersect(atac, wa=True, u=True).sort()\n",
        "fg_polr2a = polr2a_unique .intersect(atac, wa=True, u=True).sort()\n",
        "bg = atac.subtract(ctcf_unique, A =True).subtract(ep300_unique, A =True).subtract(polr2a_unique, A = True).sort()\n",
        "\n",
        "# # Intersection checks\n",
        "assert fg_ctcf.intersect(bg).total_coverage() == 0\n",
        "assert fg_ep300.intersect(bg).total_coverage() == 0\n",
        "assert fg_polr2a.intersect(bg).total_coverage() == 0"
      ],
      "metadata": {
        "id": "GtsBXeutViJh"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "[1] Create a histogram showing the distribution of region sizes (check seminar).\n"
      ],
      "metadata": {
        "id": "mYRHbC4MZ6pE"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import seaborn as sns\n",
        "import matplotlib.pyplot as plt\n",
        "\n",
        "# Raw data\n",
        "fig_raw, axes_raw = plt.subplots(1, 4, figsize=(24, 6))\n",
        "raw_configs = [\n",
        "    (\"CTCF\", ctcf, 25, axes_raw[0]),\n",
        "    (\"EP300\", ep300, 25, axes_raw[1]),\n",
        "    (\"POLR2A\", polr2a, 25, axes_raw[2]),\n",
        "    (\"ATAC-seq\", atac, 100, axes_raw[3]),\n",
        "]\n",
        "\n",
        "for title, regions, bw, ax in raw_configs:\n",
        "    sns.histplot([x.length for x in regions], binwidth=bw, kde=True, ax=ax)\n",
        "    ax.set(title=title, xlabel=\"Peak length\", ylabel=\"Density\")\n",
        "\n",
        "fig_raw.suptitle(\"Peak Length Distribution: Raw Data\", fontsize=16)\n",
        "plt.tight_layout(rect=[0, 0, 1, 0.95])\n",
        "plt.show()\n",
        "\n",
        "# Processed Data\n",
        "fig_processed, axes_processed = plt.subplots(1, 4, figsize=(24, 6))\n",
        "processed_configs = [\n",
        "    (\"CTCF & ATAC-seq\", fg_ctcf, 25, axes_processed[0]),\n",
        "    (\"EP300 & ATAC-seq\", fg_ep300, 25, axes_processed[1]),\n",
        "    (\"POLR2A & ATAC-seq\", fg_polr2a, 25, axes_processed[2]),\n",
        "    (\"ATAC-seq\", bg, 100, axes_processed[3]),\n",
        "]\n",
        "\n",
        "for title, regions, bw, ax in processed_configs:\n",
        "    sns.histplot([x.length for x in regions], binwidth=bw, kde=True, ax=ax)\n",
        "    ax.set(title=title, xlabel=\"Peak length\", ylabel=\"Density\")\n",
        "\n",
        "fig_processed.suptitle(\"Peak Length Distribution: Processed Data\", fontsize=16)\n",
        "\n"
      ],
      "metadata": {
        "id": "sR1rOg4rddQo"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "fg_dict = {\n",
        "    \"CTCF\": fg_ctcf,\n",
        "    \"POLR2A\": fg_polr2a,\n",
        "    \"EP300\": fg_ep300,\n",
        "}\n",
        "from collections import Counter\n",
        "\n",
        "def analyze_lengths_multiple(fg_dict, top_n=10):\n",
        "    for name, fg in fg_dict.items():\n",
        "        lengths = [x.length for x in fg]\n",
        "        cnts = Counter(lengths)\n",
        "        total = sum(cnts.values())\n",
        "        cnts = sorted(cnts.items(), key=lambda x: x[1], reverse=True)\n",
        "\n",
        "        print(f\"\\nАнализ набора: {name}\")\n",
        "        for k, v in cnts[:top_n]:\n",
        "            print(f\"{k} -> {v} ({v / total * 100:.2f}%)\")\n",
        "analyze_lengths_multiple(fg_dict)"
      ],
      "metadata": {
        "id": "Ezgc00-rdd-6"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## DNA sequence"
      ],
      "metadata": {
        "id": "OAN3Rmi3fqhr"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "To distinguish between our classes, we'll be using DNA-based features, so let's first extract the DNA sequence for each region:"
      ],
      "metadata": {
        "id": "qqizwYMrf6wA"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Download the genome from the Google Cloud\n",
        "!gsutil -m cp \\\n",
        "  \"gs://genomics-public-data/resources/broad/hg38/v0/Homo_sapiens_assembly38.fasta.fai\" \\\n",
        "  \"gs://genomics-public-data/resources/broad/hg38/v0/Homo_sapiens_assembly38.fasta\" \\\n",
        "  ."
      ],
      "metadata": {
        "id": "I_rSrcQsgJ4I"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "fasta = \"Homo_sapiens_assembly38.fasta\"\n",
        "\n",
        "# Fetch target sequences from the genome\n",
        "# seqfn - path to final FASTA file\n",
        "fg_ctcf_seq = fg_ctcf.sequence(fi=fasta).seqfn\n",
        "fg_ep300_seq = fg_ep300.sequence(fi=fasta).seqfn\n",
        "fg_polr2a_seq = fg_polr2a.sequence(fi=fasta).seqfn\n",
        "bgseq = bg.sequence(fi=fasta).seqfn"
      ],
      "metadata": {
        "id": "Cw2PaqdQgNN2"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!pip3 install biopython"
      ],
      "metadata": {
        "id": "tG3IZ75WhZQT"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "from Bio import SeqIO\n",
        "\n",
        "# When parsing FASTA biopython returns a Seq object,\n",
        "# the sequence itself is in the .seq field.\n",
        "fg_ctcf_seq = [str(x.seq) for x in SeqIO.parse(fg_ctcf_seq, format='fasta')]\n",
        "fg_ep300_seq = [str(x.seq) for x in SeqIO.parse(fg_ep300_seq, format='fasta')]\n",
        "fg_polr2a_seq = [str(x.seq) for x in SeqIO.parse(fg_polr2a_seq, format='fasta')]\n",
        "bgseq = [str(x.seq) for x in SeqIO.parse(bgseq, format='fasta')]\n",
        "\n",
        "# Sanity check\n",
        "print(\"Foreground example\", fg_ep300_seq[0])"
      ],
      "metadata": {
        "id": "U3gVPYUaiKqe"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import numpy as np\n",
        "\n",
        "print(f\"Before \\t fg_ctcf: {len(fg_ctcf_seq)}; fg_polr2a: {len(fg_polr2a_seq)}; fg_ep300: {len(fg_ep300_seq)}; bg: {len(bgseq)}\")\n",
        "\n",
        "np.random.seed(124)\n",
        "\n",
        "# Для fg_ctcf_seq, fg_polr2a_seq, fg_ep300_seq\n",
        "fg_ctcf_seq = np.random.choice(fg_ctcf_seq, size=2_000, replace=False)\n",
        "fg_polr2a_seq = np.random.choice(fg_polr2a_seq, size=2000, replace=False)\n",
        "fg_ep300_seq = np.random.choice(fg_ep300_seq, size=2_000, replace=False)\n",
        "\n",
        "bgseq = np.random.choice(bgseq, size=8_000, replace=False)\n",
        "\n",
        "# Проверим размер после изменений\n",
        "print(f\"After \\t fg_ctcf: {len(fg_ctcf_seq)}; fg_polr2a: {len(fg_polr2a_seq)}; fg_ep300: {len(fg_ep300_seq)}; bg: {len(bgseq)}\")\n"
      ],
      "metadata": {
        "id": "Ws_hh6y0lhfy"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "from collections import defaultdict\n",
        "\n",
        "# Inefficient but simple implementation\n",
        "def calculate_kmers(seq: str, klen: int):\n",
        "  if len(seq) < klen:\n",
        "        return {}  # Если последовательность слишком короткая, возвращаем пустой словарь\n",
        "  total_kmers = len(seq) - klen + 1\n",
        "\n",
        "  counts = defaultdict(int)\n",
        "  for ind in range(total_kmers):\n",
        "    window = seq[ind:ind+klen]\n",
        "    counts[window] += 1\n",
        "\n",
        "  # Exclude non-ATGC k-mers\n",
        "  counts = {\n",
        "      k: v for k, v in counts.items() if {\"A\", \"C\", \"G\", \"T\"}.issuperset(set(k))\n",
        "  }\n",
        "\n",
        "  # Calculate frequencies\n",
        "  total_kmers = sum(counts.values())\n",
        "  frequencies = {k: v / total_kmers for k, v in counts.items()}\n",
        "  return frequencies\n"
      ],
      "metadata": {
        "id": "DL11-H11nzIK"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "from tqdm import tqdm\n",
        "import pandas as pd\n",
        "\n",
        "# List of k-measure lengths\n",
        "KMERS = 1, 2, 3, 4, 5\n",
        "df = []\n",
        "seq_dict = {\n",
        "    0: bgseq,                # background sequences\n",
        "    1: fg_ctcf_seq,          # CTCF sequences\n",
        "    2: fg_ep300_seq,         # EP300 sequences\n",
        "    3: fg_polr2a_seq          # polr2a sequences\n",
        "}\n",
        "for cls, sequences in seq_dict.items():\n",
        "    for seq in tqdm(sequences, desc=f\"Processing TF {cls}\"):\n",
        "        record = {}\n",
        "        for klen in KMERS:\n",
        "\n",
        "\n",
        "            # Generation of k-measures for the current sequence\n",
        "            record.update(calculate_kmers(seq, klen))\n",
        "        record['Class'] = cls\n",
        "        df.append(record)\n"
      ],
      "metadata": {
        "id": "VFToQVtmmi3U"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "\n",
        "df = pd.DataFrame(df).fillna(0)\n",
        "df.head()\n",
        "# Saving dataset\n",
        "df.to_csv('transcription_factors_kmer.csv', index=False)"
      ],
      "metadata": {
        "id": "cxYO7mnSrF1r"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "[1.5] Calculate two tables showing overlaps between all experiments: before (basically with raw regions) and after calculating intersection/subtractions. The tables should look like this:"
      ],
      "metadata": {
        "id": "Fj3yegZKmlEQ"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "dirty_data = {\"CTCF\": ctcf, \"EP300\": ep300, \"POLR2A\": polr2a, \"ATAC\": atac}\n",
        "clean_data = {\"CTCF_unique\": ctcf_unique , \"EP300_unique\": ep300_unique, \"POLR2A_unique\": polr2a_unique, \"ATAC\": atac}\n",
        "\n",
        "def calculate_overlap_fraction(bed_a, bed_b):\n",
        "    intersection = bed_a.intersect(bed_b, wa=True, u=True)\n",
        "    overlap_fraction = len(intersection) / len(bed_a) if len(bed_a) > 0 else 0\n",
        "    return overlap_fraction\n",
        "\n",
        "def build_overlap_matrix(experiments):\n",
        "    matrix = pd.DataFrame(index=experiments.keys(), columns=experiments.keys())\n",
        "    for exp_a, bed_a in experiments.items():\n",
        "        for exp_b, bed_b in experiments.items():\n",
        "            matrix.loc[exp_a, exp_b] = calculate_overlap_fraction(bed_a, bed_b)\n",
        "    return matrix.astype(float)\n",
        "raw_matrix = build_overlap_matrix(dirty_data)\n",
        "processed_matrix = build_overlap_matrix(clean_data)\n",
        "print(\"Raw Overlap Matrix (Before):\")\n",
        "print(raw_matrix)\n",
        "print(\"\\nProcessed Overlap Matrix (After):\")\n",
        "print(processed_matrix)\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "_HEy34S6slN7"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "<img src=\"https://drive.google.com/uc?export=view&id=1mbGgAcLagrgIuYhkEST0Uo-duIfAW6oh\" width=\"250\"/>"
      ],
      "metadata": {
        "id": "HMJDPkLMDg-p"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "(You don't need to make it identical to this picture. Only make sure to plot the legend and labels.)"
      ],
      "metadata": {
        "id": "63Am7LaHDW8T"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Self-check here!\n",
        "\n",
        "You should expect no overlap between TFs. Also, each TF should intersect with ATAC-seq regions. For more info, read the guide for this homework."
      ],
      "metadata": {
        "id": "5gTYj0cikABW"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Sanity check: Ensure no overlaps between TFs\n",
        "assert ctcf_unique.intersect(ep300_unique).total_coverage() == 0\n",
        "assert ep300_unique.intersect(polr2a_unique).total_coverage() == 0\n",
        "assert polr2a_unique.intersect(ctcf_unique).total_coverage() == 0\n",
        "\n",
        "assert ctcf_unique.intersect(atac).total_coverage() != 0\n",
        "assert ep300_unique.intersect(atac).total_coverage() != 0\n",
        "assert polr2a_unique.intersect(atac).total_coverage() != 0"
      ],
      "metadata": {
        "id": "VfUt1X6Z2ddc"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}